### PR TITLE
WIP Add lifecycle for S3 #559

### PR DESCRIPTION
**Pull request by scotwk**
Issue #559 

Add a lifecycle action to S3.

### CODE REVIEW

**Comment by scotwk** [comment]
@kapilt - I'm having trouble getting better coverage in the tests because I can't find parameter combinations for all the parameters - I keep getting this error when I try to add in more config knobs to the lifecycle:

```
An error occurred (MalformedXML) when calling the PutBucketLifecycleConfiguration operation: The XML you provided was not well-formed or did not validate against our published schema
```

As noted in the original issue, providing a valid configuration is an exercise left to the policy writer, so if you are happy with this then it is good to go.

One question though - should we provide a corresponding `delete-lifecycle` action?

File: c7n/resources/s3.py [hunk 1]
```diff
@@ -1726,3 +1727,110 @@ def process_chunk(self, batch, bucket):
...
+                    id: default-retention-policy
+                    expire-days: 60
+                    expire-obj-del-marker: false
+                    noncur-expire-days: 60
+    """
+
+    schema = type_schema(
+        'lifecycle',
+        **{
+            'id': {'type': 'string'},
```
Comment (kapilt) [inline c7n/resources/s3.py]: if this is required should pass in required keyword set to list of keys that must be defined.

File: c7n/resources/s3.py [hunk 2]
```diff
@@ -1726,3 +1727,110 @@ def process_chunk(self, batch, bucket):
...
+            key = 'NoncurrentVersionTransitions'
+            if key not in rule:
+                rule[key] = [{}]
+            rule[key][0]['StorageClass'] = self.data['noncur-trans-type']
+
+        config = {'Rules': [rule]}
+
+        for bucket in buckets:
+            s3 = bucket_client(local_session(self.manager.session_factory), bucket)
+            s3.put_bucket_lifecycle_configuration(
```
Comment (kapilt) [inline c7n/resources/s3.py]: the new lifecycle config rule needs to be merged into extant lifecycle, this effectively replaces an extant unconditionally, when the goal would be to just replace the named rule id.

**Comment by scotwk** [comment]
Closing this without a merge as it is superceded by #1635

**Pull request closed without merging**